### PR TITLE
Improved blankposting

### PR DESCRIPTION
**Pull request by Crystalwarrior**
- Fix blankposting not hiding the chatbox (2.8.5 regression)
- Internally presume " " is the same as "" and use "" as the blankpost check (if the server sends us empty message string and not one with a single space bar for example)
- Fix screenshake not functioning with blankposting properly
- Allow user to send blankposts for the purpose of changing their emote by pressing enter without typing any message (not even spacebar), but only if the last char ID speaking matches our character ID
- Do not create spammy log entries for blankposting - we only need to know the user blankposted one time. Any subsequent blankposts are treated as simple "emote change requests" or w/e you wanna call em.

### CODE REVIEW

File: src/courtroom.cpp [hunk 1]
```diff
@@ -1590,7 +1594,10 @@ void Courtroom::on_chat_return_pressed()
 
   packet_contents.append(ao_app->get_emote(current_char, current_emote));
 
-  packet_contents.append(ui_ic_chat_message->text());
+  QString text = ui_ic_chat_message->text();
+  if (text == "")
+    text = " "; // blankpost time
```
Comment (oldmud0) [inline src/courtroom.cpp:1599]: Can we not fix this on the server side... \*head hitting wall\*

File: src/courtroom.cpp [hunk 2]
```diff
@@ -1870,24 +1877,34 @@ void Courtroom::handle_chatmessage(QStringList *p_contents)
   if (f_char_id >= 0)
     f_charname = ao_app->get_showname(char_list.at(f_char_id).name);
 
-  chatlogpiece *temp =
-      new chatlogpiece(f_charname, f_displayname, m_chatmessage[MESSAGE], false,
-                       m_chatmessage[TEXT_COLOR].toInt());
-  ic_chatlog_history.append(*temp);
-  if (ao_app->get_auto_logging_enabled())
-    ao_app->append_to_file(temp->get_full(), ao_app->log_filename, true);
+  if (m_chatmessage[MESSAGE] == " ") // User-created blankpost
```
Comment (oldmud0) [inline src/courtroom.cpp:1880]: Use [`QString::trimmed()`](https://doc.qt.io/qt-5/qstring.html#trimmed) in case user tried to use multiple spaces (e.g. old client not allowing same message to be sent twice, so they try to blankpost by adding more spaces):

```suggestion
  if (m_chatmessage[MESSAGE].trimmed().isEmpty()) // User-created blankpost
```

File: src/courtroom.cpp [hunk 3]
```diff
@@ -1590,7 +1594,10 @@ void Courtroom::on_chat_return_pressed()
 
   packet_contents.append(ao_app->get_emote(current_char, current_emote));
 
-  packet_contents.append(ui_ic_chat_message->text());
+  QString text = ui_ic_chat_message->text();
```
Comment (scatterflower) [inline src/courtroom.cpp:1597]: use a more descriptive variable name, maybe f_ictext or something

File: src/courtroom.cpp [hunk 4]
```diff
@@ -1870,24 +1877,34 @@ void Courtroom::handle_chatmessage(QStringList *p_contents)
...
-                       m_chatmessage[TEXT_COLOR].toInt());
-  ic_chatlog_history.append(*temp);
-  if (ao_app->get_auto_logging_enabled())
-    ao_app->append_to_file(temp->get_full(), ao_app->log_filename, true);
+  if (m_chatmessage[MESSAGE] == " ") // User-created blankpost
+  {
+    m_chatmessage[MESSAGE] = ""; // Turn it into true blankpost
+  }
+  // General blankpost check
+  chatmessage_is_empty = m_chatmessage[MESSAGE] == "";
```
Comment (oldmud0) [inline src/courtroom.cpp:1885]: Is this a _global_? What? Horrible! See if you can remove this so that the check is just `m_chatmessage[MESSAGE].isEmpty()`.

File: src/courtroom.cpp (hunk 2 above)
Comment (scatterflower) [inline src/courtroom.cpp:1880]: use `.trimmed().isEmpty()`

File: src/courtroom.cpp [hunk 5]
```diff
@@ -1870,24 +1877,34 @@ void Courtroom::handle_chatmessage(QStringList *p_contents)
...
+
+  if (!chatmessage_is_empty || ic_chatlog_history.isEmpty() || ic_chatlog_history.last().get_message() != "")
+  {
+    chatlogpiece *temp =
+        new chatlogpiece(f_charname, f_displayname, m_chatmessage[MESSAGE], false,
+                         m_chatmessage[TEXT_COLOR].toInt());
+    ic_chatlog_history.append(*temp);
```
Comment (oldmud0) [inline src/courtroom.cpp:1892]: Frankly never understood why `new` was used here. You should be able to add to a vector without calling `new`:

```suggestion
    chatlogpiece log_entry(f_charname, f_displayname, m_chatmessage[MESSAGE], false,
                           m_chatmessage[TEXT_COLOR].toInt());
    ic_chatlog_history.append(log_entry);
```

File: src/courtroom.cpp [hunk 6]
```diff
@@ -2763,6 +2780,10 @@ void Courtroom::start_chat_ticking()
     this->do_flash();
     sfx_player->play(ao_app->get_custom_realization(m_chatmessage[CHAR_NAME]));
   }
+  int emote_mod = m_chatmessage[EMOTE_MOD].toInt(); // text meme bonanza
```
Comment (scatterflower) [inline src/courtroom.cpp:2783]: why was this moved?

File: src/courtroom.cpp [hunk 7]
```diff
@@ -1870,24 +1877,34 @@ void Courtroom::handle_chatmessage(QStringList *p_contents)
...
+  chatmessage_is_empty = m_chatmessage[MESSAGE] == "";
+
+  if (!chatmessage_is_empty || ic_chatlog_history.isEmpty() || ic_chatlog_history.last().get_message() != "")
+  {
+    chatlogpiece *temp =
+        new chatlogpiece(f_charname, f_displayname, m_chatmessage[MESSAGE], false,
+                         m_chatmessage[TEXT_COLOR].toInt());
+    ic_chatlog_history.append(*temp);
+    if (ao_app->get_auto_logging_enabled())
+      ao_app->append_to_file(temp->get_full(), ao_app->log_filename, true);
```
Comment (oldmud0) [inline src/courtroom.cpp:1894]: Then:
```suggestion
      ao_app->append_to_file(log_entry.get_full(), ao_app->log_filename, true);
```

File: src/courtroom.cpp [hunk 8]
```diff
@@ -1538,7 +1538,11 @@ void Courtroom::append_server_chatmessage(QString p_name, QString p_message,
 
 void Courtroom::on_chat_return_pressed()
 {
-  if (ui_ic_chat_message->text() == "" || is_muted)
+  if (is_muted)
+    return;
+
+  // You can only blankpost w/o any input if the last character who spoke is yours
+  if (ui_ic_chat_message->text() == "" && m_chatmessage[CHAR_ID].toInt() != m_cid)
```
Comment (oldmud0) [inline src/courtroom.cpp:1545]: So then people will just add a space. Right?

File: src/courtroom.cpp (hunk 2 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1880]: You sure? People might want behavior for silent lip flaps with long periods of space (whispering for example)

File: src/courtroom.cpp (hunk 6 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:2783]: Consistency, variable not necessary for log parsing until the spot it was moved to

File: src/courtroom.cpp (hunk 8 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1545]: Only for their first message. Subsequent messages don't need a space

File: src/courtroom.cpp (hunk 1 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1599]: Oh lol we could

File: src/courtroom.cpp (hunk 2 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1880]: What if they wanna emulate whispering with silent lip flaps tho

File: src/courtroom.cpp (hunk 4 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1885]: Can do

File: src/courtroom.cpp (hunk 5 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1892]: Fuck accidentally resolved unresolve pls

Comment (Crystalwarrior) [reply src/courtroom.cpp:1892]: Can do

File: src/courtroom.cpp (hunk 7 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1894]: Yee

File: src/courtroom.cpp (hunk 2 above)
Comment (oldmud0) [reply src/courtroom.cpp:1880]: Players should use ellipses for that use case; e.g.
```
...        ...         ...
```

File: src/courtroom.cpp (hunk 8 above)
Comment (oldmud0) [reply src/courtroom.cpp:1545]: Why do we require a space for the first message?

Comment (Crystalwarrior) [reply src/courtroom.cpp:1545]: Just because you're asking that now we won't

File: src/courtroom.cpp (hunk 1 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1599]: actually no the packet doesn't get properly sent unless I do this for some fucking reason @oldmud0, might need further investigation but for now this works

File: src/courtroom.cpp (hunk 2 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:1880]: Good idea

File: src/courtroom.cpp (hunk 6 above)
Comment (Crystalwarrior) [reply src/courtroom.cpp:2783]: Wait oh fuckin it's because if chatmessage is empty then the screenshake won't fire without moving this @scatterflower

**Comment by scatterflower** [comment]
lgtm

**Comment by Crystalwarrior** [comment]
all that's left is @oldmud0 reviwing and merging this
